pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,1074 @@
+    #[pallet::storage]
+    #[pallet::getter(fn gas_reserve)]
+    pub(super) type GasReserve<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn min_deposit)]
+    pub(super) type MinDeposit<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn min_withdraw)]
+    pub(super) type MinWithdraw<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    // Dynamic Storage Items
+
+    /// NOTE: sum of all user's deposits, to ensure balance never drops below
+    #[pallet::storage]
+    #[pallet::getter(fn sum_of_deposits)]
+    pub(super) type SumOfDeposits<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    /// NOTE: the total pot is the total number of KMA eligible to win in the current drawing cycle
+    #[pallet::storage]
+    #[pallet::getter(fn total_pot)]
+    pub(super) type TotalPot<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn total_users)]
+    pub(super) type TotalUsers<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn is_rebalancing)]
+    pub(super) type RebalanceInProgress<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn active_balance_per_user)]
+    pub(super) type ActiveBalancePerUser<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn unclaimed_winnings_by_account)]
+    pub(super) type UnclaimedWinningsByAccount<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, BalanceOf<T>, OptionQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn total_unclaimed_winnings)]
+    pub(super) type TotalUnclaimedWinnings<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[derive(Clone, Encode, Decode, TypeInfo)]
+    pub(super) struct UnstakingCollator<AccountId, BlockNumber> {
+        pub account: AccountId,
+        pub since: BlockNumber,
+    }
+
+    #[pallet::storage]
+    pub(super) type UnstakingCollators<T: Config> =
+        StorageValue<_, Vec<UnstakingCollator<T::AccountId, T::BlockNumber>>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn remaining_unstaking_balance)]
+    pub(super) type RemainingUnstakingBalance<T: Config> =
+        StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[derive(Clone, Debug, Eq, PartialEq, Encode, Decode, TypeInfo)]
+    pub struct Request<AccountId, BlockNumber, Balance> {
+        pub user: AccountId,
+        pub block: BlockNumber,
+        pub balance: Balance,
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn withdrawal_request_queue)]
+    pub(super) type WithdrawalRequestQueue<T: Config> =
+        StorageValue<_, Vec<Request<T::AccountId, T::BlockNumber, BalanceOf<T>>>, ValueQuery>;
```

Comment:
These requests are time-bounded because they will be paid out after timelock expiry, no way for users to prevent that other than by stopping the lottery or if drawings fail to execute.
The anti-spam bound on this is `min_withdraw` again so it needs to be reasonably set

Reply:
oof yeah of course, a user requests a withdraw which fills up this queue, and it gets freed up after however many blocks, so it can't grow infinitely.